### PR TITLE
fixed a couple things with get_directory()

### PR DESCRIPTION
**Pull request by time-killer-games**
- Robert seemed to not know how to set the default directory, so I took care of that for him.
- He wrote it originally with a bug regarding the flags which allowed selecting non-directory locations that returned CLSID's instead of paths, which I also fixed by preventing those from being selectable, forcing only valid paths to be selectable.
- Fixed some things with Zenity and KDialog that Josh asked me to do in my previous pull request but he merged it before I could do so, (he said I didn't have to do it, but I wanted to).
- Changed the order of which the error and stack is shown with show_error() in all widget systems.

Edited for the sake of being kind.

### CODE REVIEW

**Comment by time-killer-games** [comment]
@RobertBColton  hey if Josh doesn't review this soon any chance if you could?

**Comment by time-killer-games** [comment]
@JoshDreamland all checks have passed. This one is ready for you to review.

**Comment by time-killer-games** [comment]
@RobertBColton @JoshDreamland are one of you stinkers gonna review this?

**Comment by time-killer-games** [comment]
@RobertBColton I'm getting the vibe you are either ignoring me or don't find this pr that important enough to be a priority to merge. I'm hoping it's the latter and not the former.

**Comment by fundies** [comment]
only im allowed to call @RobertBColton an idiot

**Comment by time-killer-games** [comment]
#bromance

**Comment by time-killer-games** [comment]
@JoshDreamland hey pooky u merge?

**Comment by time-killer-games** [comment]
@RobertBColton Josh wants you to review this. Quick sucking your thumb/penis and get this shit merged.

**Comment by time-killer-games** [comment]
@RobertBColton when I called you an idiot I was JOKING. You and fundies fight, bark, and chew on each other like wolves but all of a sudden when I partake in being silly like that it is wrong. lol lighten up man.

**Comment by RobertBColton** [comment]
I will take a look at the pull request shortly now that you've cleaned up the pull request description. We didn't get to it because fundies flooded with the MSVC prs all at once, and while trivial and positive, they are time consuming to go through.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp [hunk 1]
```diff
@@ -55,9 +55,9 @@ static string gs_but1, gs_but2, gs_but3;
 void show_error(string errortext, const bool fatal)
 {
   #ifdef DEBUG_MODE
-    errortext += enigma::debug_scope::GetErrors();
+  errortext = enigma::debug_scope::GetErrors() + "\r\n\r\n" + errortext;
```
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: This is not related to `get_directory` and should be split or explained.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp [hunk 2]
```diff
@@ -558,37 +558,51 @@ int get_color(int defcolor, bool advanced)
...
-//because I could not find out which one it uses, since IFileDialog is used by both wxWidgets and QtFramework
-//and there doesn't appear to be a standard file picker for XP or lower in the Windows API except SHBrowseForFolder that is
-//used by Game Maker for get_directory_alt
-  IFileDialog* fileDialog;
-  CoCreateInstance(CLSID_FileOpenDialog, NULL, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(&fileDialog));
+  //NOTE: This uses the Windows Vista or later file chooser, which is different than the one used by GM8 and lower
+  //because I could not find out which one it uses, since IFileDialog is used by both wxWidgets and QtFramework
+  //and there doesn't appear to be a standard file picker for XP or lower in the Windows API except SHBrowseForFolder that is
+  //used by Game Maker for get_directory_alt
+  IFileDialog *SelectDirectory;
```
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Please don't use CamelCase for variable names, those are reserved for types and sometimes functions. I don't know any style guide that recommends CamelCase variable names, so please don't do that.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp [hunk 3]
```diff
@@ -558,37 +558,51 @@ int get_color(int defcolor, bool advanced)
...
+  CoCreateInstance(CLSID_FileOpenDialog, NULL, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(&SelectDirectory));
 
   DWORD options;
-  fileDialog->GetOptions(&options);
-  options &= ~FOS_FILEMUSTEXIST;
-  options &= ~FOS_PATHMUSTEXIST;
```
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Are you sure GM doesn't allow the path or file to not exist? This seems like I would have tested this when in 32f553b4f2c07f31e359845b2b71f7c5a518696b I originally added these flags.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 1 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: I did this because it is a change made in my other pr and i didn't want any conflcts as I intend to get the other one merged first so no changes get undone.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 2 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Your original code used camel case, but i can fix that for you as well.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 3 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Yes, I am positive. No, you must have not tested properly. Did you really think a normal GM user would benefit from a CLSID that non-paths return?

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 2 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Ok, I'll change it.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 1 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: That being said, I'll let you know when Josh merges my other pr before this one is ready to be.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 2 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: I meant don't use _upper_ CamelCase for variable names, lower camelCase is totally fine as is snake case in the context of variable names. Upper CamelCase is acceptable for types like VertexBuffer. snake_case is the only one generally ok for user functions or variables. Sometimes Upper CamelCase is acceptable for class member functions/fundies likes it and used it in `EnigmaPlugin.hpp` for emake.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 3 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: No, I didn't, because I didn't know what a CLSID is because it genuinely sounds like Microsoft nonsense. I have never seen a file dialog that "selects COM objects" and while the idea doesn't sound far fetched it's totally insane. I blame them on poor API design more than myself not knowing what CLSID was. You have to understand the Microsoft document was complete garbage back in 2013 and didn't even provide code examples. The information you are finding now is easier for you because Microsoft has changed and is making a concerted effort to provide easily accessible Win32 documentation and such.

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: The only reason I discovered the CLSID was from my testing, had nothing to do with reading the docs.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 1 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Ok, that's fine, but your concerns are unwarranted. Git simply doesn't work like that. The commit just records "inserted these lines at this position" and would raise a merge conflict whenever two commits tried to insert to the same place. Just in general, unless you had a reason to include this here (maybe because the other stuff was dependent), I don't think it's a good idea to do this and it may just end up causing you more grief than it's worth. But, it's fine, I am not going to notice if the other one is merged.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 2 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Resolved: 7f1f03547f166239450aa1b2f067fc1c59f37ecd

You fixed it.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 1 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Thank you. And sorry for the flood of retarded messages on hangouts. It's not even worth reading at this point.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 3 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Regardless, it was still the best code at the time because nobody had made any attempt to add the dialog until I did. Let me back up for a second though and clarify that I am not actually asking about the CLSID. Here's the source of confusion and maybe why I removed these options originally.

I ran this little test in GM8.1 just now.
```gml
show_message(get_directory("bleh"));
show_message(get_directory_alt("ok", "bleh"));
show_message(get_open_filename("", "bleh"));
```

The third file selecting dialog I think may have been implemented in ENIGMA. In ENIGMA and GM8.1 it allows you to select files and paths that don't exist by typing the name in the file dialog and just hitting open. As you can see below, using GM8.1's open file dialog I am able to open a file that doesn't exist.

GM8 Open File Dialog Nonexistant

That does not hold true for GM8.1's get directory functions. Both of the dialogs seem outdated compared to the file open dialog and do not allow you to actually type in the name of the dialog. Correct me if I am wrong, but I don't believe ENIGMA is using these outdated dialogs for selecting directories except in one case, I believe we use the newer variant? If that's true, and the user is able to type the name of the directory in, then I think it should allow selecting directories which don't exist to be kept consistent with the open file dialog. However, that does not mean I think it should be able to select COM objects.

I don't have anything else to add to this without more information I mentioned above.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 1 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Sure, just stay on topic is all I ask, I'm very stressed right now. Thanks.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 3 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: The thing is, if you allow selecting non existent locations, certain locations that do not count as valid directories will return a CLSID, no way around it. In the case of the file dialogs, I find it a bit silly you can select non-existent files, that defeats the purpose of selecting an file to begin with; the file must exist in order to be opened and processed. You sure you aren't testing the save dialog instead?

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp [hunk 4]
```diff
@@ -558,37 +558,51 @@ int get_color(int defcolor, bool advanced)
...
+    LPWSTR szName = nullptr;
+    hr = pItem->GetDisplayName(SIGDN_NORMALDISPLAY, &szName);
+    if (SUCCEEDED(hr)) {
+      selectDirectory->SetFolder(pItem);
+      ::CoTaskMemFree(szName);
+    }
+    pItem->Release();
+  }
+
+  if (caption.empty()) selectDirectory->SetTitle(L"Select Directory");
```
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Does GM really do this if you don't set the caption? 🤕 Either way I don't like it, just curious though, and later on functions could be provided to allow the user to override the default caption. Something like `set_directory_default_caption` or something, I don't know.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp [hunk 5]
```diff
@@ -558,37 +558,51 @@ int get_color(int defcolor, bool advanced)
...
-    std::wstring wstr = wideres;
-    res = string(wstr.begin(), wstr.end());
+  selectDirectory->GetOptions(&options);
+  selectDirectory->SetOptions(options | FOS_PICKFOLDERS | FOS_NOCHANGEDIR | FOS_FORCEFILESYSTEM);
+
+  tstring tstr_dname = widen(dname);
+  LPWSTR szFilePath = (wchar_t *)tstr_dname.c_str();
+
+  IShellItem *pItem = nullptr;
+  HRESULT hr = ::SHCreateItemFromParsingName(szFilePath, nullptr, IID_PPV_ARGS(&pItem));
```
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Uh oh, do you know how new this API is? Seems to be minimum supported client is Windows Vista.
https://docs.microsoft.com/en-us/windows/desktop/api/shobjidl_core/nf-shobjidl_core-shcreateitemfromparsingname

 I don't know where you stand on this but I like to keep this system backwards compatible for Windows XP users. Yes, I know that a small part of Rusky just died right now from me saying this, but I digress. Please double check this. If the API used by my code was supported on Windows XP, please change this in some way to make it Windows XP compatible.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 4 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: It actually does that on its own, but the default caption it has it "Select Folder". I changed it to "Select Directory" because that is the caption GM's function uses.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 5 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Nevermind, I see it was Windows Vista before too.
https://docs.microsoft.com/en-us/windows/desktop/api/shobjidl_core/nn-shobjidl_core-ifiledialog

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 4 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Did you prefer it to say "Select Folder"? I can remove the excess code if you do.

Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Oh, well, I guess that's fine. But I don't know if I would have bothered because of my comment above. If we made a fallback for Windows XP (not asking you to do this just to be clear) then we'd probably use the same API GM8.1 did for its outdated get directory dialogs. In that case, it would have that title. I would have just left the newer fallback use the one Windows picks, because, if theoretically YoYoGames ever decides to bring this back, that's likely what caption it would have. So just in general I wouldn't have bothered, but it's whatever.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp [hunk 6]
```diff
@@ -558,37 +558,51 @@ int get_color(int defcolor, bool advanced)
...
+  pItem = nullptr;
+  hr = selectDirectory->GetResult(&pItem);
+
+  if (SUCCEEDED(hr)) {
+    LPWSTR wstr_result;
+    pItem->GetDisplayName(SIGDN_DESKTOPABSOLUTEPARSING, &wstr_result);
+    pItem->Release();
+
+    string str_result;
+    str_result = string_replace_all(shorten(wstr_result) + "\\", "\\\\", "\\");
```
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Why are we doing this again? I forgot about why we were doing these escapes?

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 4 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Well, the API dialog used for the original get_directory function is Delphi-specific. It only exists in Delphi applications, I know this from using Delphi and doing research on the matter. The best we could do is make an imitation dialog at best, like the one I made in the past for Dialog Module that I ended up replacing.

**Comment by RobertBColton** [comment]
Ok, I think I picked this apart good enough and know what's being changed. Very nice work figuring that out because it's a really convoluted API on Microsoft's part!

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 6 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: It's to add the final slash at the end. This is new behavior in GMStudio for directory related stuff. If the slash is added to a drive root, there would be a double slash. The string_replace_All removes the duplicate in such case to return a prettier string.

Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Where do we stand on that for the other functions? Are the other functions in Win32 doing that yet? If not, please don't do it for now. I can't remember because I thought it died in review. I'd rather that decision be made all at once to all of the dialogs, but in general I can't remember what the situation was with the Win32 widgets.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 4 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Fair enough, let's just continue with the Windows standard one because that's likely what GMS would have if it ever brought back the function.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 6 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: The only function in Win32 widgets remaining that doesn't do this is get_directory_alt which I plan to also make that change to in a different pr because it just so happens that function you wrote is also broken (the root argument doesn't work).

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: That being said, all of our other widget systems include the last slash so imo it would be best to make these write-once, compile-anywhere.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 3 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Like I said you can try it yourself in 8.1. I will say that's a good distinction you made though. If, hypothetically, it's not a bug in GM then YoYo's logic may have been that it was intended so you could select a file you want to create. But that doesn't make sense, since that's what the save dialog would be for.
```cpp
show_message(get_open_filename("", "bleh"));
```

Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Oh, I just realized, this is not resolved. I want you to consider that there is no such thing as a `get_save_directory` @time-killer-games, so how would you let your user select a directory that they want you to create? For example, what if it was about selecting a directory where you would like your game saved, and the game creates the directory if it doesn't exist? If you don't like it being able to select non-existant directories, we can make it a parameter and give it a default argument where it doesn't let you. Then we give people options. Again, I won't force you to do it, but now I see a reason to.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 6 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Is `get_open_filename` doing this? I can't seem to see where it does it.

Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Then please create a helper function for it now like you did in the other places: https://github.com/enigma-dev/enigma-dev/pull/1493/files#r245889217

That way if somebody wants they can easily change it in just one spot. Honestly the same helper function could have been used across systems here but that's not too much of a biggy.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 3 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Actually, it is. You can right click -> new folder in the dialog just like you can with get_directory_alt and the file dialogs.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 6 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Actually, there does seem to be an existing helper in there for this, please just reuse it.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 3 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Yeah I was aware of that. I guess it really doesn't matter then. There's no reason to put in the extra work to see if the CLSID came from the file system. I guess we're good on this.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 6 above)
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: I forgot about it, will do.

Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Resolved: 52b8ae95e91073b650573928c1f6d11f8f2ea7ec

You fixed it.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 4 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Resolved: 61383af99eeee11165935b158a52158258144e6c

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 1 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Leaving this one unresolved until you talk to Josh about that other pull request.

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Sounds good.

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: @RobertBColton alright, Josh mudged my other pull request, so i guess that means you can mudge this one too :) Hoorah for mudging!!!

Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Two things:

1) I disagree with the way GM8 showed error messages. Usually the error is shown first and the stack is shown second in an error message dialog because stacks can be very huge. It causes the error messages to get buried, and quite frankly, I don't really even think we need to add the extra text for run mode either.
2) This follows from above, the debug scope `GetErrors` actually has a bug. It starts with an extra newline when it doesn't need to. Also it prints the stack from top to bottom, not top to bottom. So because of (1) the stack now looks backwards. This is an error message I show in a script, but the stack looks like it occurred in the object create event.
image

In my opinion, this change should just be undone and `GetErrors` should be changed to put the new line on the end of the line rather than the first character.

**Comment by RobertBColton** [comment]
Ok, I tested `show_message(get_directory("holymoly", "captions"));` which seems to work ok. The issues with the error dialog need fixed or reverted and split out into a separate pull before I will merge.

File: ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp (hunk 1 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Here's what I think it should look like:
ENIGMA Suggested Stack Trace Dialog

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: @RobertBColton Sure, will do. Would you like me to reflect these changes in our other widgets systems or should that be a separate pr?

Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Up to you, here's a couple of things to mention. I am not entirely sure my way is perfect, just objectively good.
* I copied the above look from Python actually.
* I guess you could say showing the error stack from start to end first (like <GM8) is maybe easier for novices to understand? In the same way update events are named "step", the novice user can think of it like stepping through the program to see where there game had an error, then finally they see the error. That seems straight forward to a non-computer scientist to me (I mean we read code like a recipe top to bottom).
* However, printing a stack bottom to top like GM8 did actually requires `GetErrors` to use a second temporary stack to unwind and then build the errors string. You know, `std::stack` doesn't even allow sequential access.
* Finally, printing a stack like GM8 did is super inconsistent with a lot of stuff and I'm afraid it'll be confusing. Python prints it most recent frame last like GM8. But you know GDB's backtrace command prints it most recent frame first. Java also prints it most recent frame first. To be honest though, a lot of IDEs let you just change the sorting order of a stack trace in the UI, so it's not too big of a deal.

Anyway, to answer your question, that's basically up to you. If you agree with me here, then yes. I think the argument is there that what I've suggested is objectively better than GM8.

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: I agree, so yeah I'll update the other systems.

Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Here's debug scope's location if you're looking for it:
https://github.com/enigma-dev/enigma-dev/blob/b074d6bf06fd7aaea9d4a137a5a0c4f7bf2226f9/ENIGMAsystem/SHELL/Universal_System/debugscope.cpp#L41

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp]: Thanks!

**Comment by time-killer-games** [comment]
@RobertBColton I updated the description to include all changes made in the pr.

**Comment by time-killer-games** [comment]
@RobertBColton alright, this one is ready to mudge, when you get the chance. thanks!
